File: Capítulo_9/Recursão/Sequência de Fibonacci/Fibonacci com memoização/Código em Python/fibonacciMemoization.py
def fibonacci_memoization():
    memo = [0, 1]
    
    def fibonacci(n):
        if n < len(memo):
            return memo[n]
        # Calcula o valor, armazena no cache e retorna
        while len(memo) <= n:
            memo.append(memo[-1] + memo[-2])
        return memo[n]
    
    return fibonacci

# Exemplo de uso
fibonacci = fibonacci_memoization()
print(fibonacci(10))  # Saída: 55
print(fibonacci(20))  # Saída: 6765
print(fibonacci(30))  # Saída: 832040

File: Capítulo_9/Recursão/Calculando o fatorial de um número/Sequência de Fibonacci/Fibonacci com memoização/fibonacciMemoization.js
// O código a seguir representa a função fibonacci escrita com memoização:
function fibonacciMenoization(n) {
    const memo = [0, 1];
    const fibonacci = (n) => {
        if (memo[n] != null) return memo[n];
        return memo[n] = fibonacci(n - 1, memo) + fibonacci(n - 2, memo);
    };
    return fibonacci;
}

const fibonacci =  fibonacciMenoization();

// Exemplo de uso
console.log(fibonacci(10)); // Saída: 55
console.log(fibonacci(20)); // Saída: 6765
console.log(fibonacci(30)); // Saída: 832040